[Diagnostics] Correctly identify location of requirement failure

xedin opened the pull request:
Previously in situations like:

```swift
protocol P {}

struct S<T: P> {
  var value: T
}

_ = S(value: 42)
```

Diagnostic has reported a problem as related to "reference" to `init`
but the failing generic type requirement belongs to `S`, so a
better diagnostic in such case should mention `generic struct S`.

In order to do that we need to first de-duplicate generic type requirement 
checking since constraint system would record generic parameter multiple
times: 1. While resolving `S` and 2. When determining a type of constructor member.

And afterwards add special logic to `RequirementFailure` to determine where
generic signature came from.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Deduplicate generic requirements after merging type variables

This comes up, for example, when `Self` type parameter is bound to its
argument in a member reference, both argument and parameter can refer
to exact same requirements, that have to be deduplicated otherwise
they are just going to get "fixed" N times, which complicates diagnostics.

xedin committed:
[ConstraintSystem] Track all generic requirements fixed along current path

Such tracking makes it easier to ignore already "fixed" requirements
which have been recorded in the constraint system multiple times e.g.
a call to initializer would open both base type and initializer
method which have shared (if not the same) requirements.

xedin committed:
[Diagnostics] Correctly identify location of requirement failure

Previously in situations like:

```swift
protocol P {}

struct S<T: P> {
  var value: T
}

_ = S(value: 42)
```

Diagnostic has reported a problem as related to "reference" to `init`
but the failing generic type requirement belongs to `S`, so a
better diagnostic in such case should mention `generic struct S`.

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test

DougGregor commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -126,6 +126,33 @@ void ConstraintSystem::mergeEquivalenceClasses(TypeVariableType *typeVar1,
+  // Let's try to de-duplicate any generic requirements
+  // associated with given type variable, if it represents
+  // a generic parameter.
+  if (typeVar1->getImpl().getGenericParameter()) {
```
Should we be looking at all type variables in the equivalence class of `typeVar1` to see if any of them have a corresponding generic parameter?

xedin replied:
I'm not actually sure what do you mean. I think the API doesn't have the best name here, I'll add `isGenericParameter` to make it clearer that this type variable is supposed to represent a particular generic parameter.

DougGregor replied:
We might have two type variables `$T1` and `$T2` that are always equivalent to each other, but only one of them has an associated generic parameter. If `typeVar1` ends up being `$T1` but `$T2` is the generic parameter, we would skip this code. Is that the right behavior?

xedin replied:
Interesting thought! As I understand current behavior, we only merge type variables representing generic parameters together as a result of `matchDeepEqualityTypes` so both sides have to be generic parameters. For argument/parameter conversions we wait until at least one side is resolved, same for `BindParam`. This check is not strictly necessary I just wanted to make sure that we don't call `gatherConstraints` after every merge. Maybe I should just do `getRepresentative` if check whether it's a generic parameter.

xedin replied:
I will also try to make this work cache of "fixed" requirements I have.

xedin replied:
Looks like de-duplication is not longer indeed with caching of fixed requirements, so this is none issue.

DougGregor commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -126,6 +126,33 @@ void ConstraintSystem::mergeEquivalenceClasses(TypeVariableType *typeVar1,
+
+    llvm::SmallPtrSet<TypeBase *, 4> existingRequirements;
+    for (auto *req : requirements) {
+      auto constraintTy = req->getSecondType();
```
Should we defensively check here whether the constraint itself has a second type? I feel like the locator checked above could be insufficient.

xedin replied:
We could `assert` but requirement is always a binary constraint from one type to another.

DougGregor replied:
Okay.

xedin committed:
[ContraintSystem] Don't de-duplicate generic requirements

Looks like this is not really needed since fixed requirements
cache can handle such de-duplication.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test
